open_spiel/python/algorithms/efr.py

```diff
@@ -0,0 +1,1006 @@
+# Copyright 2019 DeepMind Technologies Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# Modified: 2023 James Flynn
+# Original: https://github.com/deepmind/open_spiel/blob/master/open_spiel/python/algorithms/cfr.py
+
+"""Python implementation of the extensive-form regret minimization algorithm.
+
+See: "Efficient Deviation Types and Learning for Hindsight Rationality in Extensive-Form Games",
+Morrill et al. 2021b, 
+https://arxiv.org/abs/2102.06973
+
+One iteration of EFR consists of:
+1) Compute current strategy from regrets (e.g. using Regret Matching).
+2) Compute values using the current strategy
+3) Compute regrets from these values
+
+The average policy converges to a Nash Equilibrium rather than the current policy.
+"""
+import copy
+from collections import defaultdict
+import attr
+
+import numpy as np
+from scipy.linalg import lstsq
+
+import pyspiel
+from open_spiel.python import policy
+
+
+@attr.s
+class _InfoStateNode(object):
+  """An object wrapping values associated to an information state."""
+  # The list of the legal actions.
```

Comment:
Blank line needed after comment (these will all be caught by the linter with Google python style)